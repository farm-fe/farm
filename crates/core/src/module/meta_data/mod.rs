use std::collections::HashMap;

use css::CssModuleMetaData;
use custom::CustomMetaDataMap;
use farmfe_macro_cache_item::cache_item;
use html::HtmlModuleMetaData;
use script::ScriptModuleMetaData;

use crate::Cacheable;

pub mod css;
pub mod custom;
pub mod html;
pub mod script;

/// Module meta data shared by core plugins through the compilation
/// Meta data which is not shared by core plugins should be stored in [ModuleMetaData::Custom]
#[cache_item]
pub enum ModuleMetaData {
  Script(Box<ScriptModuleMetaData>),
  Css(Box<CssModuleMetaData>),
  Html(Box<HtmlModuleMetaData>),
  Custom(CustomMetaDataMap),
}

impl Default for ModuleMetaData {
  fn default() -> Self {
    Self::Custom(CustomMetaDataMap::default())
  }
}

impl ToString for ModuleMetaData {
  fn to_string(&self) -> String {
    match self {
      Self::Script(_) => "script".to_string(),
      Self::Css(_) => "css".to_string(),
      Self::Html(_) => "html".to_string(),
      Self::Custom(_) => "custom".to_string(),
    }
  }
}

impl Clone for ModuleMetaData {
  fn clone(&self) -> Self {
    match self {
      Self::Script(script) => Self::Script(script.clone()),
      Self::Css(css) => Self::Css(css.clone()),
      Self::Html(html) => Self::Html(html.clone()),
      Self::Custom(custom) => {
        let mut custom_new = HashMap::default();
        for (k, v) in custom.iter() {
          let cloned_data = v.serialize_bytes().unwrap();
          let cloned_custom = v.deserialize_bytes(cloned_data).unwrap();
          custom_new.insert(k.clone(), cloned_custom);
        }
        Self::Custom(CustomMetaDataMap::from(custom_new))
      }
    }
  }
}

impl ModuleMetaData {
  pub fn as_script_mut(&mut self) -> &mut ScriptModuleMetaData {
    if let Self::Script(script) = self {
      script
    } else {
      panic!("ModuleMetaData is not Script")
    }
  }

  pub fn as_script(&self) -> &ScriptModuleMetaData {
    if let Self::Script(script) = self {
      script
    } else {
      panic!("ModuleMetaData is not Script but {:?}", self.to_string())
    }
  }

  pub fn as_css(&self) -> &CssModuleMetaData {
    if let Self::Css(css) = self {
      css
    } else {
      panic!("ModuleMetaData is not css")
    }
  }

  pub fn as_css_mut(&mut self) -> &mut CssModuleMetaData {
    if let Self::Css(css) = self {
      css
    } else {
      panic!("ModuleMetaData is not css")
    }
  }

  pub fn as_html(&self) -> &HtmlModuleMetaData {
    if let Self::Html(html) = self {
      html
    } else {
      panic!("ModuleMetaData is not html")
    }
  }

  pub fn as_html_mut(&mut self) -> &mut HtmlModuleMetaData {
    if let Self::Html(html) = self {
      html
    } else {
      panic!("ModuleMetaData is not html")
    }
  }

  /// get custom meta data by key
  pub fn get_custom_mut<T: Cacheable + Default>(&mut self, key: &str) -> &mut T {
    if let Self::Custom(custom) = self {
      custom.get_mut(key).unwrap()
    } else {
      panic!("ModuleMetaData is not Custom")
    }
  }
}
