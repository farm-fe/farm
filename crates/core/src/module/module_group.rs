use std::collections::{HashMap, HashSet};

use petgraph::{
  stable_graph::{DefaultIx, NodeIndex, StableDiGraph},
  visit::{Bfs, Dfs, DfsPostOrder, EdgeRef, IntoEdgeReferences},
};

use crate::resource::{resource_pot::ResourcePotId, resource_pot_map::ResourcePotMap};

use super::{module_graph::ModuleGraph, ModuleId};

/// A `entry_module_id -> ModuleGroup` map
#[derive(Debug)]
pub struct ModuleGroupGraph {
  /// internal graph
  g: StableDiGraph<ModuleGroup, ()>,
  /// to index module in the graph using [ModuleId]
  id_index_map: HashMap<ModuleId, NodeIndex<DefaultIx>>,
}

impl ModuleGroupGraph {
  pub fn new() -> Self {
    Self {
      g: StableDiGraph::new(),
      id_index_map: HashMap::new(),
    }
  }

  pub fn replace(&mut self, other: ModuleGroupGraph) {
    self.g = other.g;
    self.id_index_map = other.id_index_map;
  }

  pub fn add_module_group(&mut self, module_group: ModuleGroup) {
    let module_group_id = module_group.id.clone();

    if self.has(&module_group_id) {
      panic!("module group already exists: {:?}", module_group_id);
    }

    let node_index = self.g.add_node(module_group);
    self.id_index_map.insert(module_group_id, node_index);
  }

  pub fn add_edge(&mut self, from: &ModuleId, to: &ModuleId) {
    let from_node_index = self.id_index_map.get(from).unwrap();
    let to_node_index = self.id_index_map.get(to).unwrap();

    self.g.add_edge(*from_node_index, *to_node_index, ());
  }

  pub fn remove_edge(&mut self, from: &ModuleId, to: &ModuleId) {
    let from_node_index = self.id_index_map.get(from).unwrap_or_else(|| {
      panic!(
        "ModuleGroupGraph::remove_edge: from {} to {}. Not found: {}",
        from.to_string(),
        to.to_string(),
        from.to_string()
      )
    });
    let to_node_index = self.id_index_map.get(to).unwrap_or_else(|| {
      panic!(
        "ModuleGroupGraph::remove_edge: from {} to {}. Not found: {}",
        from.to_string(),
        to.to_string(),
        to.to_string()
      )
    });
    let edge_index = self.g.find_edge(*from_node_index, *to_node_index).unwrap();
    self.g.remove_edge(edge_index);
  }

  pub fn remove_module_group(&mut self, id: &ModuleGroupId) -> Option<ModuleGroup> {
    let node_index = self.id_index_map.remove(id).unwrap();
    self.g.remove_node(node_index)
  }

  pub fn module_group(&self, id: &ModuleGroupId) -> Option<&ModuleGroup> {
    let node_index = self.id_index_map.get(id).unwrap();
    self.g.node_weight(*node_index)
  }

  pub fn module_group_mut(&mut self, id: &ModuleGroupId) -> Option<&mut ModuleGroup> {
    let node_index = self.id_index_map.get(id)?;

    self.g.node_weight_mut(*node_index)
  }

  /// get the topologically sorted module groups
  pub fn module_groups(&self) -> Vec<&ModuleGroup> {
    self.g.node_weights().collect()
  }

  /// the same as [ModuleGroupGraph::module_groups], but mutable.
  pub fn module_groups_mut(&mut self) -> Vec<&mut ModuleGroup> {
    self.g.node_weights_mut().collect()
  }

  pub fn has(&self, id: &ModuleGroupId) -> bool {
    self.id_index_map.contains_key(id)
  }

  pub fn has_edge(&self, from: &ModuleId, to: &ModuleId) -> bool {
    let from_node_index = self.id_index_map.get(from);
    let to_node_index = self.id_index_map.get(to);

    if from_node_index.is_none() || to_node_index.is_none() {
      return false;
    }

    self
      .g
      .find_edge(*from_node_index.unwrap(), *to_node_index.unwrap())
      .is_some()
  }

  pub fn len(&self) -> usize {
    self.g.node_count()
  }

  pub fn is_empty(&self) -> bool {
    self.g.node_count() == 0
  }

  pub fn dfs(&self, entry: &ModuleGroupId, op: &mut dyn FnMut(&ModuleGroupId)) {
    let mut dfs = Dfs::new(&self.g, *self.id_index_map.get(entry).unwrap());

    while let Some(node_index) = dfs.next(&self.g) {
      op(&self.g[node_index].id);
    }
  }

  pub fn dfs_post_order(&self, entry: &ModuleGroupId, op: &mut dyn FnMut(&ModuleGroupId)) {
    let mut dfs = DfsPostOrder::new(&self.g, *self.id_index_map.get(entry).unwrap());

    while let Some(node_index) = dfs.next(&self.g) {
      op(&self.g[node_index].id);
    }
  }

  pub fn bfs(&self, entry: &ModuleGroupId, op: &mut dyn FnMut(&ModuleGroupId)) {
    let mut bfs = Bfs::new(&self.g, *self.id_index_map.get(entry).unwrap());

    while let Some(node_index) = bfs.next(&self.g) {
      op(&self.g[node_index].id);
    }
  }

  pub fn dependencies(&self, module_id: &ModuleId) -> Vec<&ModuleGroup> {
    let node_index = self.id_index_map.get(module_id).unwrap();
    let mut dependencies = Vec::new();

    for edge in self.g.edges(*node_index) {
      dependencies.push(&self.g[edge.target()]);
    }

    dependencies
  }

  pub fn dependencies_ids(&self, module_id: &ModuleId) -> Vec<ModuleId> {
    let node_index = self.id_index_map.get(module_id).unwrap();
    let mut dependencies = Vec::new();

    for edge in self.g.edges(*node_index) {
      dependencies.push(self.g[edge.target()].id.clone());
    }

    dependencies
  }

  pub fn dependents(&self, module_id: &ModuleId) -> Vec<&ModuleGroup> {
    let node_index = self.id_index_map.get(module_id).unwrap();
    let mut dependents = Vec::new();

    for edge in self
      .g
      .edges_directed(*node_index, petgraph::Direction::Incoming)
    {
      dependents.push(&self.g[edge.source()]);
    }

    dependents
  }

  pub fn toposort(&self, entries: Vec<ModuleGroupId>) -> Vec<ModuleGroupId> {
    let mut sorted = Vec::new();
    let mut visited = HashSet::new();

    for entry in entries {
      self.dfs_post_order(&entry, &mut |id| {
        if !visited.contains(id) {
          sorted.push(id.clone());
          visited.insert(id.clone());
        }
      });
    }

    sorted.reverse();
    sorted
  }

  pub fn print_graph(&self) {
    println!("digraph {{\n nodes:");

    for node in self.g.node_weights() {
      println!("  \"{}\";", node.id.to_string());
    }

    println!("\nedges:");

    for edge in self.g.edge_references() {
      let source = self.g[edge.source()].id.to_string();
      let target = self.g[edge.target()].id.to_string();
      println!("  \"{}\" -> \"{}\";", source, target);
    }

    println!("}}");
  }
}

impl Default for ModuleGroupGraph {
  fn default() -> Self {
    Self::new()
  }
}

impl PartialEq for ModuleGroupGraph {
  fn eq(&self, other: &Self) -> bool {
    let mut self_module_groups = self.module_groups();
    self_module_groups.sort_by_key(|g| g.id.clone());

    let mut other_module_groups = other.module_groups();
    other_module_groups.sort_by_key(|g| g.id.clone());

    let mut self_edges = self
      .g
      .edge_references()
      .map(|e| {
        let source = self.g[e.source()].id.clone();
        let target = self.g[e.target()].id.clone();

        (source, target)
      })
      .collect::<Vec<_>>();
    self_edges.sort();

    let mut other_edges = other
      .g
      .edge_references()
      .map(|e| {
        let source = other.g[e.source()].id.clone();
        let target = other.g[e.target()].id.clone();

        (source, target)
      })
      .collect::<Vec<_>>();
    other_edges.sort();

    self_module_groups == other_module_groups && self_edges == other_edges
  }
}

impl Eq for ModuleGroupGraph {}

pub type ModuleGroupId = ModuleId;

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct ModuleGroup {
  /// the module group's id is the same as its entry module's id.
  pub id: ModuleGroupId,
  /// the modules that this group has
  modules: HashSet<ModuleId>,
  /// the [ResourcePot]s this group merged to
  resource_pots: HashSet<ResourcePotId>,
}

impl ModuleGroup {
  pub fn new(id: ModuleGroupId) -> Self {
    Self {
      modules: HashSet::from([id.clone()]),
      id,
      resource_pots: HashSet::new(),
    }
  }

  pub fn add_module(&mut self, module_id: ModuleId) {
    self.modules.insert(module_id);
  }

  pub fn remove_module(&mut self, module_id: &ModuleId) {
    self.modules.retain(|id| id != module_id);
  }

  pub fn modules(&self) -> &HashSet<ModuleId> {
    &self.modules
  }

  pub fn add_resource_pot(&mut self, resource_pot_id: ResourcePotId) {
    self.resource_pots.insert(resource_pot_id);
  }

  pub fn remove_resource_pot(&mut self, resource_pot_id: &ResourcePotId) {
    self.resource_pots.retain(|id| id != resource_pot_id);
  }

  pub fn resource_pots(&self) -> &HashSet<ResourcePotId> {
    &self.resource_pots
  }

  pub fn sorted_resource_pots(
    &self,
    module_graph: &ModuleGraph,
    resource_pot_map: &ResourcePotMap,
  ) -> Vec<ResourcePotId> {
    let mut resource_pots_order_map = HashMap::<String, usize>::new();
    let mut sorted_resource_pots = self
      .resource_pots()
      .iter()
      .cloned()
      .collect::<Vec<_>>();

    sorted_resource_pots.iter().for_each(|rp| {
      let rp = resource_pot_map.resource_pot(rp).unwrap();
      let min_order = rp
        .modules()
        .iter()
        .map(|m| {
          let module = module_graph.module(m).unwrap();
          module.execution_order
        })
        .min()
        .unwrap_or(0);

      resource_pots_order_map.insert(rp.id.to_string(), min_order);
    });

    sorted_resource_pots.sort_by(|a, b| {
      let a_order = resource_pots_order_map.get(&a.to_string()).unwrap_or(&0);
      let b_order = resource_pots_order_map.get(&b.to_string()).unwrap_or(&0);

      a_order.cmp(b_order)
    });

    sorted_resource_pots
  }

  pub fn set_resource_pots(&mut self, resource_pots: HashSet<ResourcePotId>) {
    self.resource_pots = resource_pots;
  }

  pub fn has_resource_pot(&self, resource_pot_id: &ResourcePotId) -> bool {
    self.resource_pots.contains(resource_pot_id)
  }
}
