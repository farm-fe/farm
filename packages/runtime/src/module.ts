/* eslint-disable @typescript-eslint/no-explicit-any */
export class Module {
  id: string;
  exports: any;
  meta: Record<string, any>;
  require: (id: string) => any;

  dispose?: () => void;

  constructor(id: string, require: (id: string) => any) {
    this.id = id;
    this.exports = {};
    this.meta = {
      env: {}
    };
    this.require = require;
  }

  onDispose(callback: () => void) {
    this.dispose = callback;
  }
}
