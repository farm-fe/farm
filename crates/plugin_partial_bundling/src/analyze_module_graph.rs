use farmfe_core::module::module_group::{ModuleGroupId, ModuleGroupType};
use farmfe_core::HashMap;
use farmfe_core::{
  module::{
    module_graph::ModuleGraph,
    module_group::{ModuleGroup, ModuleGroupGraph},
    ModuleId,
  },
  HashSet,
};
use std::collections::VecDeque;

pub fn module_group_graph_from_module_graph(module_graph: &mut ModuleGraph) -> ModuleGroupGraph {
  let get_entries = |entries: &HashMap<ModuleId, String>| {
    entries
      .clone()
      .into_iter()
      .map(|(entry, _)| entry)
      .collect()
  };

  let mut module_group_graph =
    module_group_graph_from_entries(&get_entries(&module_graph.entries), module_graph);
  // handle dynamic entries
  let dynamic_module_groups =
    module_groups_from_dynamic_entries(get_entries(&module_graph.dynamic_entries), module_graph);
  for group in dynamic_module_groups {
    module_group_graph.add_module_group(group);
  }

  module_group_graph
}

pub fn module_group_graph_from_entries(
  entries: &Vec<ModuleId>,
  module_graph: &mut ModuleGraph,
) -> ModuleGroupGraph {
  let mut module_group_graph = ModuleGroupGraph::new();
  let mut edges = vec![];
  let mut visited = HashSet::default();

  for entry in entries.clone() {
    if visited.contains(&entry) {
      continue;
    }

    let (group, dynamic_dependencies) =
      module_group_from_entry(&entry, ModuleGroupType::Entry, module_graph);
    edges.extend(dynamic_dependencies.clone().into_iter().map(|dep| {
      (
        group.id.clone(),
        ModuleGroupId::new(&dep, &ModuleGroupType::DynamicImport),
      )
    }));

    module_group_graph.add_module_group(group);

    visited.insert(entry);
    let mut queue = VecDeque::from(dynamic_dependencies);

    while !queue.is_empty() {
      let head = queue.pop_front().unwrap();

      if visited.contains(&head) {
        continue;
      }

      visited.insert(head.clone());

      let (group, dynamic_dependencies) =
        module_group_from_entry(&head, ModuleGroupType::DynamicImport, module_graph);
      edges.extend(dynamic_dependencies.clone().into_iter().map(|dep| {
        (
          group.id.clone(),
          ModuleGroupId::new(&dep, &ModuleGroupType::DynamicImport),
        )
      }));

      module_group_graph.add_module_group(group);
      queue.extend(dynamic_dependencies);
    }
  }

  for (from, to) in &edges {
    module_group_graph.add_edge(from, to);
  }

  module_group_graph
}

/// get module group start from a entry. return (module group, dynamic dependencies)
/// traverse the module graph using bfs, stop when reach a dynamic dependency
fn module_group_from_entry(
  entry: &ModuleId,
  module_group_type: ModuleGroupType,
  graph: &mut ModuleGraph,
) -> (ModuleGroup, Vec<ModuleId>) {
  let mut visited = HashSet::default();
  // ModuleGroupType
  let mut module_group = ModuleGroup::new(entry.clone(), module_group_type);
  let mut dynamic_entries = vec![];

  graph
    .module_mut(entry)
    .unwrap()
    .module_groups
    .insert(module_group.id.clone());

  visited.insert(entry.clone());

  let deps = graph
    .dependencies(entry)
    .into_iter()
    .map(|(k, v)| (k, v.is_dynamic_import(), v.is_dynamic_entry()))
    .collect::<Vec<_>>();

  for (dep, is_dynamic_import, is_dynamic_entry) in deps {
    if is_dynamic_import {
      if !dynamic_entries.contains(&dep) {
        dynamic_entries.push(dep);
      }
    } else if !is_dynamic_entry {
      // visited all dep and its dependencies using BFS
      let mut queue = VecDeque::new();
      queue.push_back(dep.clone());

      while !queue.is_empty() {
        let head = queue.pop_front().unwrap();

        if visited.contains(&head) {
          continue;
        }

        visited.insert(head.clone());
        module_group.add_module(head.clone());
        graph
          .module_mut(&head)
          .unwrap()
          .module_groups
          .insert(module_group.id.clone());

        for (dep, edge) in graph.dependencies(&head) {
          if edge.is_dynamic_import() {
            if !dynamic_entries.contains(&dep) {
              dynamic_entries.push(dep);
            }
          } else if !edge.is_dynamic_entry() {
            queue.push_back(dep);
          }
        }
      }
    }
  }

  (module_group, dynamic_entries)
}

pub fn module_groups_from_dynamic_entries(
  dynamic_entries: Vec<ModuleId>,
  module_graph: &mut ModuleGraph,
) -> Vec<ModuleGroup> {
  let mut module_groups = vec![];

  for entry in dynamic_entries {
    let mut group = ModuleGroup::new(entry.clone(), ModuleGroupType::DynamicEntry);
    // find al deep dependencies of dynamic entries using BFS
    let mut queue = VecDeque::from([entry.clone()]);
    let mut visited = HashSet::default();

    while !queue.is_empty() {
      let head = queue.pop_front().unwrap();

      if visited.contains(&head) {
        continue;
      }

      visited.insert(head.clone());
      group.add_module(head.clone());
      module_graph
        .module_mut(&head)
        .unwrap_or_else(|| panic!("module {:?} not found", head))
        .module_groups
        .insert(group.id.clone());

      queue.extend(module_graph.dependencies_ids(&head));
    }

    module_groups.push(group.clone());
  }

  module_groups
}

#[cfg(test)]
mod tests {
  use std::sync::Arc;

  use farmfe_core::{
    context::CompilationContext,
    module::module_group::{ModuleGroupId, ModuleGroupType},
    parking_lot::RwLock,
    plugin::{Plugin, PluginHookContext, ResolveKind},
    HashMap, HashSet,
  };
  #[cfg(test)]
  use farmfe_testing_helpers::construct_test_module_graph;
  use farmfe_testing_helpers::{
    assert_debug_snapshot, construct_test_module_graph_complex, construct_test_module_group_graph,
  };

  use crate::FarmPluginPartialBundling;

  use super::module_group_from_entry as mgfe;

  #[test]
  fn analyze_module_graph() {
    let plugin = FarmPluginPartialBundling {
      partial_bundling_groups_enforce_map: Default::default(),
    };
    let mut context = CompilationContext::new(Default::default(), vec![]).unwrap();
    let graph = construct_test_module_graph();

    let _ = std::mem::replace(&mut context.module_graph, Box::new(RwLock::new(graph)));
    let context = Arc::new(context);
    let mut module_graph = context.module_graph.write();

    let module_group_graph = plugin
      .analyze_module_graph(
        &mut module_graph,
        &context,
        &PluginHookContext {
          caller: None,
          meta: HashMap::default(),
        },
      )
      .unwrap()
      .unwrap();

    assert_eq!(module_group_graph.len(), 5);
    let group_id_a = ModuleGroupId::new(&"A".into(), &ModuleGroupType::Entry);
    let group_id_b = ModuleGroupId::new(&"B".into(), &ModuleGroupType::Entry);
    let group_id_d = ModuleGroupId::new(&"D".into(), &ModuleGroupType::DynamicImport);
    let group_id_f = ModuleGroupId::new(&"F".into(), &ModuleGroupType::DynamicImport);
    let group_id_g = ModuleGroupId::new(&"G".into(), &ModuleGroupType::DynamicImport);
    assert!(module_group_graph.has(&group_id_a));
    assert!(module_group_graph.has(&group_id_b));
    assert!(module_group_graph.has(&group_id_d));
    assert!(module_group_graph.has(&group_id_f));
    assert!(module_group_graph.has(&group_id_g));

    let module_group_a = module_group_graph.module_group(&group_id_a).unwrap();
    assert_eq!(module_group_a.entry_module_id, "A".into());
    assert_eq!(
      module_group_a.modules(),
      &HashSet::from_iter(["A".into(), "C".into()])
    );

    let module_group_b = module_group_graph.module_group(&group_id_b).unwrap();
    assert_eq!(module_group_b.entry_module_id, "B".into());
    assert_eq!(
      module_group_b.modules(),
      &HashSet::from_iter(["B".into(), "D".into(), "E".into()])
    );

    let module_group_d = module_group_graph.module_group(&group_id_d).unwrap();
    assert_eq!(module_group_d.entry_module_id, "D".into());
    assert_eq!(module_group_d.modules(), &HashSet::from_iter(["D".into()]));

    let module_group_f = module_group_graph.module_group(&group_id_f).unwrap();
    assert_eq!(module_group_f.entry_module_id, "F".into());
    assert_eq!(
      module_group_f.modules(),
      &HashSet::from_iter(["F".into(), "A".into(), "C".into()])
    );

    let module_group_g = module_group_graph.module_group(&group_id_g).unwrap();
    assert_eq!(module_group_g.entry_module_id, "G".into());
    assert_eq!(module_group_g.modules(), &HashSet::from_iter(["G".into()]));

    let test_pairs = vec![
      ("A", vec![&group_id_a, &group_id_f]),
      ("B", vec![&group_id_b]),
      ("C", vec![&group_id_a, &group_id_f]),
      ("D", vec![&group_id_d, &group_id_b]),
      ("E", vec![&group_id_b]),
      ("F", vec![&group_id_f]),
      ("G", vec![&group_id_g]),
    ];

    for tp in test_pairs {
      let m_a = module_graph.module_mut(&tp.0.into()).unwrap();
      assert_eq!(m_a.module_groups.len(), tp.1.len());

      for g_id in tp.1 {
        assert!(m_a.module_groups.contains(g_id));
      }
    }
  }

  #[test]
  fn module_group_from_entry() {
    let mut graph = construct_test_module_graph();

    let (module_group, de) = mgfe(&"A".into(), ModuleGroupType::Entry, &mut graph);
    assert_eq!(de, vec!["F".into(), "D".into()]);
    assert_eq!(module_group.entry_module_id, "A".into());
    assert_eq!(
      module_group.modules(),
      &HashSet::from_iter(["A".into(), "C".into()])
    );
    assert!(graph
      .module(&"A".into())
      .unwrap()
      .module_groups
      .contains(&module_group.id));
    assert!(graph
      .module(&"C".into())
      .unwrap()
      .module_groups
      .contains(&module_group.id));
  }

  #[test]
  fn module_group_graph_from_entries() {
    let mut graph = construct_test_module_graph();

    let entries = vec!["A".into(), "B".into()];
    let module_group_graph = crate::module_group_graph_from_entries(&entries, &mut graph);
    let final_module_group_graph = construct_test_module_group_graph();

    assert_eq!(module_group_graph, final_module_group_graph);
  }

  #[test]
  fn module_groups_from_dynamic_entries() {
    let mut graph = construct_test_module_graph_complex();

    let dynamic_entries = vec!["A".into(), "E".into()];
    let module_groups = super::module_groups_from_dynamic_entries(dynamic_entries, &mut graph);

    let mut module_group_entry_module_ids = module_groups
      .iter()
      .map(|mg| mg.entry_module_id.clone())
      .collect::<Vec<_>>();
    module_group_entry_module_ids.sort();
    assert_eq!(module_group_entry_module_ids, vec!["A".into(), "E".into()]);

    let test_pairs = vec![
      ("A", vec!["A", "C", "D", "F", "H"]),
      ("E", vec!["E", "G", "H"]),
    ];

    assert_eq!(module_groups.len(), test_pairs.len());

    for (mg_id, modules) in test_pairs {
      let mg = module_groups
        .iter()
        .find(|mg| mg.entry_module_id == mg_id.into())
        .unwrap();
      let mut mg_modules = mg
        .modules()
        .iter()
        .map(|m| m.to_string())
        .collect::<Vec<_>>();
      mg_modules.sort();
      assert_eq!(mg_modules, modules);

      for m in modules {
        assert!(graph
          .module(&m.into())
          .unwrap()
          .module_groups
          .contains(&mg.id));
      }
    }
  }

  #[test]
  fn module_group_graph_from_module_graph() {
    let mut graph = construct_test_module_graph_complex();
    // update A -> D to dynamic entry
    graph
      .update_edge(
        &"A".into(),
        &"D".into(),
        farmfe_core::module::module_graph::ModuleGraphEdge::new(vec![
          farmfe_core::module::module_graph::ModuleGraphEdgeDataItem {
            kind: ResolveKind::DynamicEntry {
              name: "AD".to_string(),
              output_filename: None,
            },
            ..Default::default()
          },
        ]),
      )
      .unwrap();
    graph.dynamic_entries = HashMap::from_iter([("D".into(), "D".to_string())]);

    let module_group_graph = crate::module_group_graph_from_module_graph(&mut graph);
    assert_debug_snapshot!(module_group_graph);
  }
}
