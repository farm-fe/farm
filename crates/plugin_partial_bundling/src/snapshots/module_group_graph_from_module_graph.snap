---
source: crates/plugin_partial_bundling/src/analyze_module_graph.rs
assertion_line: 400
input_file: crates/testing_helpers/src/lib.rs
---
ModuleGroupGraph {
    g: StableGraph {
        Ty: "Directed",
        node_count: 5,
        edge_count: 4,
        edges: (0, 1), (1, 1), (2, 1), (2, 3),
        node weights: {
            0: ModuleGroup {
                id: ModuleGroupId(
                    "A_Entry",
                ),
                entry_module_id: ModuleId {
                    relative_path: "A",
                    query_string: "",
                },
                module_group_type: Entry,
                modules: {
                    ModuleId {
                        relative_path: "C",
                        query_string: "",
                    },
                    ModuleId {
                        relative_path: "A",
                        query_string: "",
                    },
                },
                resource_pots: {},
            },
            1: ModuleGroup {
                id: ModuleGroupId(
                    "F_DynamicImport",
                ),
                entry_module_id: ModuleId {
                    relative_path: "F",
                    query_string: "",
                },
                module_group_type: DynamicImport,
                modules: {
                    ModuleId {
                        relative_path: "C",
                        query_string: "",
                    },
                    ModuleId {
                        relative_path: "A",
                        query_string: "",
                    },
                    ModuleId {
                        relative_path: "F",
                        query_string: "",
                    },
                    ModuleId {
                        relative_path: "H",
                        query_string: "",
                    },
                },
                resource_pots: {},
            },
            2: ModuleGroup {
                id: ModuleGroupId(
                    "B_Entry",
                ),
                entry_module_id: ModuleId {
                    relative_path: "B",
                    query_string: "",
                },
                module_group_type: Entry,
                modules: {
                    ModuleId {
                        relative_path: "D",
                        query_string: "",
                    },
                    ModuleId {
                        relative_path: "E",
                        query_string: "",
                    },
                    ModuleId {
                        relative_path: "B",
                        query_string: "",
                    },
                    ModuleId {
                        relative_path: "H",
                        query_string: "",
                    },
                },
                resource_pots: {},
            },
            3: ModuleGroup {
                id: ModuleGroupId(
                    "G_DynamicImport",
                ),
                entry_module_id: ModuleId {
                    relative_path: "G",
                    query_string: "",
                },
                module_group_type: DynamicImport,
                modules: {
                    ModuleId {
                        relative_path: "H",
                        query_string: "",
                    },
                    ModuleId {
                        relative_path: "G",
                        query_string: "",
                    },
                },
                resource_pots: {},
            },
            4: ModuleGroup {
                id: ModuleGroupId(
                    "D_DynamicEntry",
                ),
                entry_module_id: ModuleId {
                    relative_path: "D",
                    query_string: "",
                },
                module_group_type: DynamicEntry,
                modules: {
                    ModuleId {
                        relative_path: "D",
                        query_string: "",
                    },
                    ModuleId {
                        relative_path: "C",
                        query_string: "",
                    },
                    ModuleId {
                        relative_path: "A",
                        query_string: "",
                    },
                    ModuleId {
                        relative_path: "F",
                        query_string: "",
                    },
                    ModuleId {
                        relative_path: "H",
                        query_string: "",
                    },
                },
                resource_pots: {},
            },
        },
        free_node: NodeIndex(4294967295),
        free_edge: EdgeIndex(4294967295),
    },
    id_index_map: {
        ModuleGroupId(
            "F_DynamicImport",
        ): NodeIndex(1),
        ModuleGroupId(
            "A_Entry",
        ): NodeIndex(0),
        ModuleGroupId(
            "B_Entry",
        ): NodeIndex(2),
        ModuleGroupId(
            "D_DynamicEntry",
        ): NodeIndex(4),
        ModuleGroupId(
            "G_DynamicImport",
        ): NodeIndex(3),
    },
}
